test/test_common/utility.h

```diff
@@ -340,8 +342,51 @@ class TestHeaderMapImpl : public HeaderMapImpl {
 
 } // namespace Http
 
+namespace Stats {
+/**
+ * This is a heap test allocator that works similar to how the shared memory allocator works in
+ * terms of reference counting, etc.
+ */
+class TestAllocator : public RawStatDataAllocator {
+public:
+  ~TestAllocator() { EXPECT_TRUE(stats_.empty()); }
+
+  RawStatData* alloc(const std::string& name) override {
+    CSmartPtr<RawStatData, freeAdapter>& stat_ref = stats_[name];
+    if (!stat_ref) {
+      stat_ref.reset(static_cast<RawStatData*>(::calloc(RawStatData::size(), 1)));
+      stat_ref->initialize(name);
+    } else {
+      stat_ref->ref_count_++;
+    }
+
+    return stat_ref.get();
+  }
+
+  void free(RawStatData& data) override {
+    if (--data.ref_count_ > 0) {
+      return;
+    }
+
+    for (auto i = stats_.begin(); i != stats_.end(); i++) {
```

Comment:
This was pre-existing, but you have a map here, can't we just remove it via the name?  I don't really care about perf for the test but I think the code would just be:
```
  if (stats_.erase(std::string(data.name()) == 0) {
    FAIL();
  }
```
which is fewer lines :)